cpp/custrings/text/ngram.cu

```diff
@@ -109,3 +110,308 @@ NVStrings* NVText::create_ngrams(NVStrings& strs, unsigned int ngrams, const cha
     return NVStrings::create_from_index((std::pair<const char*,size_t>*)d_results,ngrams_count);
 }
 
+//
+using position_pair = thrust::pair<int32_t,int32_t>;
+using index_pair = thrust::pair<const char*,size_t>;
+
+// This class walks a string looking for specified delimiter character(s).
+// It will automatically ignore adjacent delimiters (i.e. different than split).
+// The next_token method returns character start position (spos) and end
+// position (epos) between delimiter runs identifying each token.
+// An iterator is used to retrieve each utf8 character to be checked.
+// The spaces parameter identifies a run of delimiters (or not delimiters).
+struct base_string_tokenize
+{
+    custring_view_array d_strings;
+    custring_view* d_delimiter;
+    custring_view* d_separator;
+
+    __device__ bool is_delimiter(Char ch)
+    {
+        if( !d_delimiter )
+            return (ch <= ' '); // all ascii whitespace
+        return d_delimiter->find(ch)>=0;
+    }
+
+    __device__ bool next_token( custring_view* dstr, bool& spaces,
+                                custring_view::iterator& itr, int& spos, int& epos )
+    {
+        if( spos >= dstr->chars_count() )
+            return false;
+        for( ; itr != dstr->end(); ++itr )
+        {
+            Char ch = *itr;
+            if( spaces == is_delimiter(ch) )
+            {
+                if( spaces )
+                    spos = itr.position()+1;
+                else
+                    epos = itr.position()+1;
+                continue;
+            }
+            spaces = !spaces;
+            if( spaces )
+            {
+                epos = itr.position();
+                break;
+            }
+        }
+        return spos < epos;
+    }
+};
+
+// per string
+struct string_token_counter_fn : base_string_tokenize
```

Comment:
Could you add some description about this class? It is obvious given this PR but it would certain help with clarity.

Reply:
Added _real_ comments to each of the new classes here.